FRW-7700 Updated project guidlines (#2808)
spryker/spryker-docs: docs/dg/dev/guidelines/project-development-guidelines.md

Also, you need to update you infrastructure and use latest provided docker images. E.g. each year we update PHP version and provide new docker images for it. Also all code updates from Spryker side will require new PHP version, so you need to update it as well.

## Apply coding guidelines
Starting from the first day of development, apply the [coding guidelines](/docs/dg/dev/guidelines/coding-guidelines/coding-guidelines.html). 
Pay attention to an [architecture convention](/docs/dg/dev/architecture/architectural-convention.html) page as in addition it provides a set of rules and recommendations that applicable specifically for project development.

## Use custom namespaces
As a recommendation, instead of using the `Pyz` namespace, you can use your own namespace for project development. For example, use the project name as a namespace. But be advised that most of existing examples and documentation use the `Pyz` namespace.

## Use custom names
Use custom names for everything that is added on the project level, like the following:
- Transfers
- Transfer properties

> - DB tables (using project namespace as a prefix is recommended)

Reviewer: @helen-laktionova don't we have a more strict rule here, considering Code Upgrader

- DB table fields,
- Modules
- Classes

> For example, customize the names by adding the project name. This will help to avoid conflicts with the Spryker core and other projects.

Reviewer: We need more strict statements here. @helen-laktionova again calling for your help.

## Avoid using, extending, and overriding Private API
Instead of using, extending, and overriding [Private API](/docs/dg/dev/architecture/module-api/declaration-of-module-apis-public-and-private.html) (everything that is not a Public API), send a request about the missing endpoints to your Spryker account manager. Spryker offers extension points that allow you to extend via the Public API and helps you to customize the application. We recommend to use the extension points instead of overriding Private API.

> Extending Private API is still possible, but core development teams are not considering changing those classes as any BC break as long as a Public API is not affected. So you can get a critical error with your next minor update.

Reviewer: ```suggestion
Extending Private API is still possible, but Spryker development team can change them without a notice in the minor change, since Spryker BC break policy is only considering a Public API.
```

> As a recommendation to avoid such issues, we recommend to not extend classes, but implement new one based on the same interface. In this case you still can get a minor update where this Private API was changed, but it much easier to fix it. This approach is not valid if you want to extend e.g. Facade, because you will be forced to re-implement all the original methods.

Reviewer: Let's rephrase this, it's very hard to read

> If there is an error, or you want to submit a fix or a new functionality by yourself, you can create a pull request to the required repository. It will be reviewed by the Spryker team and, if it is accepted, it will be merged into the core via our release process. Be advised that such PRs will be checked as a core contribution and should be prepared accordingly.

Reviewer: ```suggestion
If you believe that you found a bug in Spryker module, or you want to submit a fix, new functionality by yourself, you can create a pull request to the corresponding module repository. It will be reviewed by the Spryker team and, if it is accepted, it will be merged into the core via our release process. Be advised that such PRs will be checked as a Core contribution and should be prepared accordingly.
```

Reviewer: Can we please post a link to the Core contribution guidelines here?

## Establish coding standards
Before you start developing, establish coding standards. Implement code checks based on the standards into your CI/CD. The default code checks shipped with Spryker are located in `/config/install/sniffs.yml`. You might want to add more checks that are based on your project's requirements.

> Make sure that code can't be merged until it corresponds to your coding standards. But [Don't be slowed down by Spryker's core rules](#dont-be-slowed-down-by-sprykers-core-rules).

Reviewer: ```suggestion
Make sure that the code is merged only when it corresponds to your coding standards. But [Don't be slowed down by Spryker's core rules](#dont-be-slowed-down-by-sprykers-core-rules).
```

- [Code Style Sniffer](/docs/dg/dev/sdks/sdk/development-tools/code-sniffer.html)
- [PHP Mess detector](https://github.com/spryker/architecture-sniffer)

## Don't be slowed down by Spryker's core rules

> Spryker core development has a lot of rules that are not always applicable to the project level. If you find that a rule is slowing you down, you are not forced to use it. But pay attention that some of the rules are not just recommendations, but conventions and must be followed in order to get a working functionality.

Reviewer: Do we have a section explaining which rules are fine to be ignored?

Author: In Arch Conventions page (that is attached in Apply coding guidelines section) mention rules that are core specific only as I know

Do not use arrays as a return type if possible. Use Transfer objects instead. It will help you to keep your code a bit more strict and understandable for the future developers (or even for you in the future).

Don't leave unused code in the project. It will make your project bigger and harder to maintain. If you don't need it, remove it. Spryker tends to deprecate code that is not used, but on the project you can just remove it. If you really need it in the future, you can check the history of the file in the git.

Spryker code usually focused on code quality and keeping the code clean. We recommend to follow the same approach in your project, but from time to time this approach can decrease a performance. So you need to find a balance between the code quality and performance.

Do not leave credentials in the code.

Define Transfer objects on the project level even if it's defined in the core. It will help you avoid unnecessary dependencies and will make your code more flexible.

> Use minor lock on the project for modules that you updated and extended private API. It will help you to avoid critical errors with the next minor update.

Reviewer: ```suggestion
Use minor lock on the project for modules that you updated and extended private API. It will help you to avoid critical errors with the next minor update. Installing [Composer Constrainer)[https://github.com/spryker-sdk/composer-constrainer] and including `code:constraint:modules` helps automating this process.
```

> DO NOT use a deprecated code from Spryker. It will be eventually removed and you will need to update your code. It's better to update it now. If you think that deprecation was a mistake, you can create a pull request to the Spryker core.

Reviewer: ```suggestion
We recommend stopping usage of the deprecated code from Spryker. It will be eventually removed and you will need to update your code, thus minimizing amount of used deprecation helps project in a long run.
```
I wouldn't be so hard here, since we are still not in  good shape with deprecation, and with `no major` policy, deprecated code is living for years.
